src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -388,22 +393,50 @@ struct
                     acc)
             in
             cvars [ xs; ys; n_acc ] t
-        | EC_endoscalar { state; n8; a8; b8 } ->
+        | EC_endoscalar { state } ->
             let t = H.feed_string t "ec_endoscale_scalar" in
-            let t =
-              Array.fold state ~init:t
-                ~f:(fun
-                     acc
-                     { n0; n8; a0; b0; a8; b8; x0; x1; x2; x3; x4; x5; x6; x7 }
-                   ->
-                  cvars
-                    [ n0; n8; a0; b0; a8; b8; x0; x1; x2; x3; x4; x5; x6; x7 ]
-                    acc)
-            in
-            cvars [ n8; a8; b8 ] t )
+            Array.fold state ~init:t
+              ~f:(fun
+                   acc
+                   { n0; n8; a0; b0; a8; b8; x0; x1; x2; x3; x4; x5; x6; x7 }
+                 ->
+                cvars
+                  [ n0; n8; a0; b0; a8; b8; x0; x1; x2; x3; x4; x5; x6; x7 ]
+                  acc) )
     | _ ->
         failwith "Unsupported constraint"
 
+  (** Converts the set of permutations (equivalence_classes) to
+      a hash table that maps each position to the next one.
+      For example, if one of the equivalence class is [pos1, pos3, pos7],
+      the function will return a hashtable that maps pos1 to pos3,
+      pos3 to pos7, and pos7 to pos1 *)
+  let equivalence_classes_to_hashtbl sys =
+    let module Relative_position = struct
+      module T = struct
+        type t = Row.t Position.t [@@deriving hash, sexp, compare]
+      end
+
+      include T
+      include Core_kernel.Hashable.Make (T)
+    end in
+    let equivalence_classes = V.Table.create () in
+    Hashtbl.iteri sys.equivalence_classes ~f:(fun ~key ~data ->
+        let u = Hashtbl.find_exn sys.union_finds key in
+        Hashtbl.update equivalence_classes (Union_find.get u) ~f:(function
+          | None ->
+              Relative_position.Hash_set.of_list data
+          | Some ps ->
+              List.iter ~f:(Hash_set.add ps) data ;
+              ps)) ;
```

Comment:
if `u <> key` shouldn't you keep track of `key` and remove that entry from `sys.equivalence_classes` later? IIUC only the entry under `u` will matter at this point.